Optimize grouped Conv3d performance

allwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36355 Optimize grouped Conv3d performance**

Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

Differential Revision: [D20957073](https://our.internmc.facebook.com/intern/diff/D20957073)

Commit:
Optimize grouped Conv3d performance

Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU
- bypassing unfolding when kernel_size = 1

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias
- test_Conv_1x1

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

[ghstack-poisoned]

Commit:
Update on "Optimize grouped Conv3d performance"


Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

Differential Revision: [D20957073](https://our.internmc.facebook.com/intern/diff/D20957073)

[ghstack-poisoned]

Commit:
Update on "Optimize grouped Conv3d performance"


Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

Differential Revision: [D20957073](https://our.internmc.facebook.com/intern/diff/D20957073)

[ghstack-poisoned]

BIT-silence commented on aten/src/ATen/native/ConvolutionMM3d.cpp:
```diff
@@ -297,7 +344,25 @@ void slow_conv3d_backward_out_cpu_template(
+          {groups, weight2d.size(1), weight2d.size(0) / groups})[g] =
+          weight2d
+              .reshape({groups, weight2d.size(0) / groups, weight2d.size(1)})[g]
+              .transpose(0, 1);
```
Not sure if .transpose(0, 1).contiguous() is needed.

ngimel replied:
Looks like it's not needed, 
```
a=torch.randn(2,3,4)
b=torch.arange(24).reshape(2,4,3) # to make sure that it is copied in the correct order. 
a[0]=b[0].transpose(0,1)
```
works and produces expected results. slice of b is copied here, thus it can be discontiguous.

allwu replied:
I see, we can allow the weights to be discontiguous, right? I will update the impl.

ngimel replied:
What you have is fine, you don't have to change it, and .contiguous() is not necessary

BIT-silence commented on aten/src/ATen/native/ConvolutionMM3d.cpp:
```diff
@@ -297,7 +344,25 @@ void slow_conv3d_backward_out_cpu_template(
+    }
+    tweight2d = tweight2d.reshape({weight2d.size(1), weight2d.size(0)});
+  } else {
+    tweight2d = weight2d.transpose(0, 1);
```
ditto.

xiaomengy reviewed (approved):
LGTM

ngimel commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -829,7 +844,13 @@ std::tuple<Tensor,Tensor,Tensor> _convolution_double_backward(
+  // TODO: hacky way of inferring the groups number for grouped Conv3D
+  // See: https://github.com/pytorch/pytorch/pull/36355
+  if (!params.transposed && input.dim() > 4) {
+    params.groups = input.size(1) / weight.size(1);
```
So potentially this could mask the issue of someone putting in input with incompatible sizes? E.g. I have Conv3d(16,32) that expects input with 16 channels. I send an input with 32 channels by mistake, and what'll happen here is instead of erroring out you say - ok, we'll run convolution with 2 groups. This seems unsafe?

allwu replied:
The shape will be checked in the forward pass here:
https://github.com/pytorch/pytorch/blob/ee3d046f8783e9a0a30305bc9ad264dc48061ecf/aten/src/ATen/native/Convolution.cpp#L584
As long as the user doesn't call double_backward directly we should be fine, right? We are counting on that the backward call is called by autograd thus removing the user error. Otherwise, we need to modify tools/autograd/derivatives.yaml and specify the group parameter.

ngimel replied:
Thanks for clarifying! That's ok then, user cannot call double backward directly (it's not exposed), so we should be fine.

ngimel reviewed (commented):
Using bmm instead of mm for grouped convolution is a clever solution, and generally, once you error-checked the input sizes, determining number of groups as input.size(1)/weight.size(1) should be fine, but I'm afraid now this hack in Convolution.cpp shadows proper input size checking.

Commit:
Update on "Optimize grouped Conv3d performance"


Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

Differential Revision: [D20957073](https://our.internmc.facebook.com/intern/diff/D20957073)

[ghstack-poisoned]

Commit:
Update on "Optimize grouped Conv3d performance"


Summary:
Resolving issue in https://github.com/pytorch/pytorch/issues/36155, by:
- supporting grouped conv3d in ```slow_conv3d```
- adding a fast path in ```__convolution``` to call ```slow_conv3d``` when
  running grouped conv3d on CPU

Test Plan:
Added the following test cases in test_nn.py, testing both forward and
backward:
- test_Conv3d_groups_nobias
- test_Conv3d_groups_wbias

Reviewers: lingyiliu

Subscribers:

Tasks: T65249457

Tags:

Differential Revision: [D20957073](https://our.internmc.facebook.com/intern/diff/D20957073)

[ghstack-poisoned]
